libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,223 @@
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/ControlFlowRevertPruner.h>
+
+#include <libsolutil/Algorithms.h>
+
+
+namespace solidity::frontend
+{
+
+namespace {
+
+/// Find the right scope for the called function: When calling a base function, we keep the most derived, but we use the called contract in case it is a library function or nullptr for a free function
+ContractDefinition const* findScopeContract(FunctionDefinition const& _function, ContractDefinition const* _callingContract)
+{
+	if (auto const* functionContract = _function.annotation().contract)
+	{
+		if (_callingContract && _callingContract->derivesFrom(*functionContract))
+			return  _callingContract;
+		else
+			return functionContract;
+	}
+
+	return nullptr;
+}
+}
+
+void ControlFlowRevertPruner::run()
+{
+	// build a lookup table for function calls / callers
+	for (auto& [pair, flow]: m_cfg.allFunctionFlows())
+		collectCalls(*pair.function, pair.contract);
+
+	findRevertStates();
+	modifyFunctionFlows();
+}
+
+FunctionDefinition const* ControlFlowRevertPruner::resolveCall(FunctionCall const& _functionCall, ContractDefinition const* _contract)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return nullptr;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	FunctionDefinition const* returnFunctionDef = &unresolvedFunctionDefinition;
+
+	if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+	{
+		if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+		{
+			if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+				if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+				{
+					solAssert(contractType->isSuper(), "");
+					ContractDefinition const* superContract = contractType->contractDefinition().superContract(*_contract);
+
+					returnFunctionDef = &unresolvedFunctionDefinition.resolveVirtual(
+						*_contract,
+						superContract
+					);
+				}
+		}
+		else
+		{
+			solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+			returnFunctionDef = &unresolvedFunctionDefinition;
+		}
+	}
+	else if (auto const* identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+	{
+		solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+		returnFunctionDef = &unresolvedFunctionDefinition.resolveVirtual(*_contract);
+	}
+
+	if (returnFunctionDef && returnFunctionDef->isImplemented())
```

Comment:
Can you move that to the call site? This is kind of an unexpected behaviour, just judging from the name of the function.

Reply:
I moved it in there to avoid repeating it at the call site every time. I can then just do 
```
					if (resolvedFunction == nullptr)
						continue;

```